packages/typir-langium/src/features/langium-type-creator.ts

```diff
@@ -0,0 +1,125 @@
+    protected readonly typeGraph: TypeGraph;
+
+    constructor(typirServices: TypirServices, langiumServices: LangiumSharedServices) {
+        this.typeGraph = typirServices.graph;
+
+        // for new and updated documents
+        langiumServices.workspace.DocumentBuilder.onBuildPhase(DocumentState.IndexedReferences, async (documents, cancelToken) => {
+            for (const document of documents) {
+                await interruptAndCheck(cancelToken);
+
+                // notify Typir about each contained node of the processed document
+                this.handleProcessedDocument(document);
+            }
+        });
+        // for deleted documents
+        langiumServices.workspace.DocumentBuilder.onUpdate((_changed, deleted) => {
+            deleted
+                .map(del => getDocumentKeyForURI(del))
+                .forEach(del => this.handleDeletedDocument(del));
+        });
+
+        // get informed about added/removed types
+        this.typeGraph.addListener(this);
+    }
+
+    abstract onInitialize(): void;
+
+    abstract onNewAstNode(domainElement: AstNode): void;
+
+    /**
+     * Starts the initialization.
+     * If this method is called multiple times, the initialization is done only once.
+     */
+    triggerInitialization() {
+        if (!this.initialized) {
+            this.onInitialize();
+            this.initialized = true;
+        }
+    }
+
+    protected handleProcessedDocument(document: LangiumDocument): void {
+        this.triggerInitialization();
+        this.currentDocumentKey = getDocumentKeyForDocument(document); // remember the key in order to map newly created types to the current document
+
+        // remove all types which were associated with the current document
+        this.handleDeletedDocument(this.currentDocumentKey);
+
+        // create all types for this document
+        AstUtils.streamAst(document.parseResult.value)
+            .forEach((node: AstNode) => this.onNewAstNode(node));
+
+        this.currentDocumentKey = '';
+    }
+
+    protected handleDeletedDocument(documentKey: string): void {
+        (this.documentTypesMap.get(documentKey) ?? [])
+            // this is the central way to remove types from the type systems, there is no need to inform the kinds
+            .forEach(typeToRemove => this.typeGraph.removeNode(typeToRemove));
+    }
+
+    addedType(newType: Type): void {
+        // the TypeGraph notifies about newly created Types
+        if (this.currentDocumentKey) {
+            // associate the new type with the current Langium document!
+            let types = this.documentTypesMap.get(this.currentDocumentKey);
+            if (!types) {
+                types = [];
+                this.documentTypesMap.set(this.currentDocumentKey, types);
+            }
+            types.push(newType);
+        } else {
+            // types which don't belong to a Langium document
+        }
+    }
+
+    removedType(_type: Type): void {
```

Comment:
Technically it would be possible to override this, but I have no use case for that in mind. I added this comment `// since this type creator actively removes types from the type graph itself, there is no need to react on removed types`.

Reply:
Sounds good.